### PR TITLE
bare bones cdc

### PR DESCRIPTION
**Pull request by tylerwilliams**

### CODE REVIEW

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 1]
```diff
@@ -1705,6 +1706,209 @@ func (z *zstdCompressor) Close() error {
...
+	if err != nil {
+		return err
+	}
+
+	if err = db.Set(keyBytes, protoBytes, &pebble.WriteOptions{Sync: false}); err == nil {
+		partitionID := md.GetFileRecord().GetIsolation().GetPartitionId()
+		p.sendSizeUpdate(partitionID, key.CacheType(), md.GetStoredSizeBytes())
+		metrics.DiskCacheAddedFileSizeBytes.With(prometheus.Labels{metrics.CacheNameLabel: p.name}).Observe(float64(md.GetStoredSizeBytes()))
+	}
+	return nil
```
Comment (sluongng) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:1909]:
```suggestion
	return err
```

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 2]
```diff
@@ -1705,6 +1706,209 @@ func (z *zstdCompressor) Close() error {
...
+		FileRecord:      fileRecord,
+		StorageMetadata: wcm.Metadata(),
+		StoredSizeBytes: int64(written),
+		LastAccessUsec:  now,
+		LastModifyUsec:  now,
+	}
+	if err := p.writeMetadata(ctx, cdcw.db, key, md); err != nil {
+		return err
+	}
+	cdcw.writtenChunks = append(cdcw.writtenChunks, r)
```
Comment (sluongng) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:1841]: I wonder if we could parallelize the lower part of  `writeRawChunk()` in an ErrGroup by moving this `writtenChunks` accumulation up the stack, inside `newCDCWriter()`. But it's probably not worth it for now 🤔 

It definitely took me a bit longer to find this line after the initial read as my intuition was telling me that this should be somewhere in the for-loop above.

**Review by sluongng — COMMENTED** [review]
An early review as I just finished reading some CDC articles.

Overall this is very exciting topic to learn about 🎉

File: deps.bzl [hunk 3]
```diff
@@ -2727,6 +2727,12 @@ def install_buildbuddy_dependencies(workspace_name = "buildbuddy"):
         sum = "h1:Ts/E8zCSEsG17dUqv7joXJFybuMLjQfWE04tsBODTxk=",
         version = "v1.0.0",
     )
+    go_repository(
+        name = "com_github_jotfs_fastcdc_go",
+        importpath = "github.com/jotfs/fastcdc-go",
+        sum = "h1:WHYIGk3k9NumGWfp4YMsemEcx/s4JKpGAa6tpCpHJOo=",
+        version = "v0.2.0",
+    )
```
[the remark this answers is not in the document]
Comment (bduffany) [reply deps.bzl:2735]: I was curious to learn about this was well so I played around a bit with the fastcdc-go impl as the rust impl --

It seems the only improvement in the 2020 paper is the "roll two bytes each time" optimization. In the rust impl I was only able to get a 1%-7% speedup on Linux x86_64 with that optimization (some discussion [here](https://github.com/nlfiedler/fastcdc-rs/issues/27) - the author got 20% speedup but using an M1 chip), and with the Go implementation I was not able to get any significant speedup at all (presumably, the inner loop in Go is less optimized than rust)

Also, one thing to note is that the rust impl performed about 10X faster, so it seems there is a lot of room for speedup if we use CGO or something like that.

[the remark this answers is not in the document]
Comment (tylerwilliams) [reply deps.bzl:2735]: haha lot of comments on a draft CL :P I love it.

There's more than just FastCDC out there btw -- here are some other (possibly faster) approaches:
 - RapidCDC: https://github.com/moking/RapidCDC
 - QuickCDC: https://github.com/jrobhoward/quickcdc
 - UltraCDC: https://ieeexplore.ieee.org/abstract/document/9894295

In general I want to get the blake3 and chunking support in first and then add CDC as an optimization on top of that. There are some considerations to whatever we pick here:
 - it ought to be easily implemented in java + go
 - should be standardized across implementations
 - should be as fast as possible or at least faster (at a minumum) than our hashing algorithm

**Comment by tylerwilliams** [comment]
implemented by @luluz66

**Pull request closed without merging**